### ISSUE TITLE
Support network event messages

### ISSUE THREAD

**Issue by linclark** [firefox-devtools/original-console-fork#195]
Parent issue: #140 

Where it is received: https://dxr.mozilla.org/mozilla-central/source/devtools/client/webconsole/webconsole.js#3330

Where it comes from: https://dxr.mozilla.org/mozilla-central/source/devtools/server/actors/webconsole.js#1522

Current tests:
- [ ] browser_console_netlogging?
- [ ] https://dxr.mozilla.org/mozilla-central/source/devtools/shared/webconsole/test/test_network_get.html
- [ ] https://dxr.mozilla.org/mozilla-central/source/devtools/shared/webconsole/test/test_network_longstring.html
- [ ] https://dxr.mozilla.org/mozilla-central/source/devtools/shared/webconsole/test/test_network_post.html
- [ ] Filtering: browser_webconsole_bug_632817.js
- [ ] Interactions: browser_webconsole_dont_navigate_on_doubleclick.js

**Comment by linclark** [comment]
I talked about this with @janodvarko and he was interested in taking it on. Honza, I'll mention you in here when it's ready. It depends on #140

**Comment by linclark** [comment]
@janodvarko After looking more closely at #140, I think it makes sense to handle everything in this issue.

Here's what I think needs to happen:
1. Change the `_onNetworkEvent` code in `webconsole.js` to something like this:

```
  _onNetworkEvent: function (type, networkInfo) {
    if (this.webConsoleFrame) {
      if (this.webConsoleFrame.NEW_CONSOLE_OUTPUT_ENABLED) {
        this.dispatchMessageAdd(networkInfo);
        return;
      }
      this.webConsoleFrame.handleNetworkEvent(networkInfo);
    }
  },
```
1. Add handling for network event packets in `utils/messages.js`. This should convert our packet structure to Chrome RDP's packet structure. <del>You can figure out the Chrome RDP structure by downloading their source, [instructions here](https://docs.google.com/document/d/1WNF-KqRSzPLUUfZqQG5AFeU_Ll8TfWYcJasa_XGf7ro/edit?pref=2&pli=1#heading=h.9t7np5fa400). Let me know if you run into trouble with this.</del> A better way to do this is probably using the [Sniffing](https://developer.chrome.com/devtools/docs/debugger-protocol) instructions.
2. Create a `NetworkEvent` component to format the message. This will need to be added in `MessageContainer` for the system to properly route the message.

**Comment by linclark** [comment]
@janodvarko given that the Chrome protocol is up in the air, let's postpone the conversion to Chrome packet structure. You can just copy the packet properties to a ConsoleMessage in utils/messages.js

FYI, if you need to add any properties to ConsoleMessage, you'll have to add it to the Immutable.Record definition in types.js

Let me know if you run into any obstacles or places where it gets confusing

**Comment by steveck-chung** [comment]
@linclark I just started a WIP in https://github.com/steveck-chung/gecko-dev/commit/a9385917206ddd231a37681d0ca00256c1e37b84 , but make network event fit into ConsoleMessage confused me. Since there would have some differences between 2 logs, such like the the stacktracer in the console message would be network status + timestamps for network message. Should we create another schema for NetworkEventMessage, or create new properties in ConsoleMessage just for network event?

**Comment by linclark** [comment]
@steveck-chung For now, lets create another one for NetworkEventMessage. Then we can decide whether to merge it back in

**Comment by steveck-chung** [comment]
Since I might not be able to finish this feature in the near future, I'll split the patch into different commits. This patch https://github.com/steveck-chung/gecko-dev/commit/a31ee47ae62ab3c320b69824c2aa8783cda6bf41 could display network event's method and url as messageText by using ConsoleMessage and ConsoleApiCall component. And I might need to create NetworkEventMessage/NetworkEvent component to distinguish the different messages(and maybe additional subcomponent for network status).

**Comment by linclark** [comment]
@steveck-chung thanks! could you submit it as a PR to this repo?

**Comment by steveck-chung** [comment]
@linclark Sure I could submit the first part. Should I add some test for this(maybe in messages.test.js since I didn't create any additional component right now )?

**Comment by linclark** [comment]
@steveck-chung Thank you for thinking of that. I'd say we can leave off the test until it's being formatted with the component.

**Comment by rickychien** [comment]
@linclark I'd like to finish following part to create a react component for network message.

Should we file a separate issue to finish this?

**Comment by linclark** [comment]
That sounds great! I can commit the PR that Steve put up as Part 1, and then you can do the rest on top of that.

**Comment by linclark** [comment]
This will land in m-c in [Bug 1300058](https://bugzilla.mozilla.org/show_bug.cgi?id=1300058)

**Comment by linclark** [comment]
@rickychien I just committed the PR from Steve, so this should be ready for work. I also did a merge of the changes from m-c. When you fetch, you'll likely need to do a full build. We only merge from m-c every two weeks or so, so this doesn't happen too often.

**Comment by rickychien** [comment]
Sure! I'm rebuilding now

**Comment by rickychien** [comment]
@linclark 
Since I'm going to create a new event object (an Immutable.Record) for NetworkEventMessage, so what's different between ConsoleMessage and NetworkEventMessage?

**Comment by linclark** [comment]
@rickychien I haven't looked closely into the data that we need for NetworkEventMessages. We are trying to match Chrome RDP for a lot of things, so we might end up with something like [this](https://chromedevtools.github.io/debugger-protocol-viewer/tot/Network/#event-responseReceived).

For now, you can make your best guess based on what data comes in via the packet and what data you need in the component.

**Comment by rickychien** [comment]
WIP preview for supporting network log.

<img width="1022" alt="screen shot 2016-09-07 at 3 54 58 pm" src="https://cloud.githubusercontent.com/assets/958165/18304174/7aaf383c-7513-11e6-9888-fa50d39f55e2.png">

Status bar and click to enable netInfoBody will be part of NetworkUpdateEventMessage. I'm going to check what else remain for NetworkEventMessage. Pull request will be submitted soon.
- [x] Method
- [x] XHR
- [x] URL
- [x] click URL and link to Network Panel
- [x] Partial support for network status (it will be fully supported in NetworkUpdateMessage)
- [x] Mixed content (Test website: https://people.mozilla.org/~tvyas/mixedboth.html)
- [x] Learn more link for [Mixed content]
- [x] click Learn more link for [Mixed content] and link to Network Panel

**Comment by rickychien** [comment]
@linclark 
I'm trying to support mixed-content style on top of message log and now we use `WebConsoleUtils.isMixedHTTPSRequest()` to check url but I've no idea how to get the url from current tab (the second parameter `this.contentLocation`).

Do you have idea how to achieve that?

**Comment by nchevobbe** [comment]
@rickychien it looks like `this.contentLocation` is assigned in `onLocationChange`, which is called in `handleTabNavigated` ( http://searchfox.org/mozilla-central/source/devtools/client/webconsole/webconsole.js#1980 ).
Maybe we could dispatch an action from there if the new console is activated, and store the `contentLocation` in the app state, so we can retrieve it easily  ?

**Comment by rickychien** [comment]
@nchevobbe cool! 
IIUC, We're going to create a locationChange action and dispatch it in ```onLocationChange``` in webconsole (http://searchfox.org/mozilla-central/source/devtools/client/webconsole/webconsole.js#1948)
However, I don't know how to invoke ```store.dispatch(uiActions.locationChange(uri))``` within webconsole.js? Is it possible to export the same store object from new-console-output-wrapper and use it in web console.js?

**Comment by rickychien** [comment]
@nchevobbe Thanks! it works

**Comment by rickychien** [comment]
Test case for network event has been added!
